File: leetcode/greatest-common-divisor-of-strings.py
# 1071. Greatest Common Divisor of Strings
# 🟢 Easy
#
# https://leetcode.com/problems/greatest-common-divisor-of-strings/
#
# Tags: Math - String

import timeit
from math import gcd


# If both input strings are formed by concatenating the same substring,
# then concatenating them should be an associative operation, we can
# use that to check if they have a GCD string. Once we know that they do,
# we know that the length of the GCD string will be the length of the
# GCD of their respective lengths.
#
# Time complexity: O(m+n) - We need to iterate both input strings to
# concatenate them and check if they have a GCD. The `gcd` function
# takes O(log(m*n)) time.
# Space complexity: O(m+n) - The concatenated strings use extra memory.
#
# Runtime 22 ms Beats 99.41%
# Memory 13.8 MB Beats 98.52%
class Solution:
    def gcdOfStrings(self, str1: str, str2: str) -> str:
        # A function to return the GCD between two numbers using the
        # Euclidean algorithm. Iterative version with O(1) space.
        def euc(a: int, b: int) -> int:
            if a < b:
                a, b = b, a
            while b:
                a, b = b, a % b
            return a

        # If the result of concatenating the strings is the same
        # independently of the order in which we concatenate them, then
        # they are both formed of the same substrings.
        return (
            str1[: gcd(len(str1), len(str2))]
            if str1 + str2 == str2 + str1
            else ""
        )


def test():
    executors = [Solution]
    tests = [
        ["LEET", "CODE", ""],
        ["ABCABC", "ABC", "ABC"],
        ["ABABAB", "ABAB", "AB"],
        ["ABABABAB", "ABAB", "ABAB"],
        ["ABBAABBA", "ABBA", "ABBA"],
    ]
    for executor in executors:
        start = timeit.default_timer()
        for _ in range(1):
            for col, t in enumerate(tests):
                sol = executor()
                result = sol.gcdOfStrings(t[0], t[1])
                exp = t[2]
                assert result == exp, (
                    f"\033[93m» {result} <> {exp}\033[91m for"
                    + f" test {col} using \033[1m{executor.__name__}"
                )
        stop = timeit.default_timer()
        used = str(round(stop - start, 5))
        cols = "{0:20}{1:10}{2:10}"
        res = cols.format(executor.__name__, used, "seconds")
        print(f"\033[92m» {res}\033[0m")


test()

File: leetcode/greatest-common-divisor-of-strings.rs
// 1071. Greatest Common Divisor of Strings
// 🟢 Easy
//
// https://leetcode.com/problems/greatest-common-divisor-of-strings/
//
// Tags: Math - String

struct Solution;
impl Solution {
    // If both input strings are formed by concatenating the same substring,
    // then concatenating them should be an associative operation, we can
    // use that to check if they have a GCD string. Once we know that they do,
    // we know that the length of the GCD string will be the length of the
    // GCD of their respective lengths.
    //
    // Time complexity: O(m+n) - We need to iterate both input strings to
    // concatenate them and check if they have a GCD. The `gcd` function
    // takes O(log(m*n)) time.
    // Space complexity: O(m+n) - The concatenated strings use extra memory.
    //
    // Runtime 0 ms Beats 100%
    // Memory 2.1 MB Beats 66.67%
    pub fn gcd_of_strings(str1: String, str2: String) -> String {
        /// A function that computes the GCD between two given non-negative
        /// integers. There is a more efficient version in Wikipedia.
        /// https://en.wikipedia.org/wiki/Binary_GCD_algorithm
        fn gcd(mut a: usize, mut b: usize) -> usize {
            if a < b {
                (a, b) = (b, a);
                // The above works on rustc 1.66 but fails in v1.58
                // return gcd(b, a);
            }
            while b != 0 {
                (a, b) = (b, a % b);
                // The above works on rustc 1.66 but fails in v1.58
                // let tmp = a % b;
                // a = b;
                // b = tmp;
            }
            a
        }
        if str1.clone() + &str2.clone() != str2.clone() + &str1.clone() {
            String::from("")
        } else {
            str1[..gcd(str1.len(), str2.len())].to_owned()
        }
    }
}

// Tests.
fn main() {
    assert_eq!(
        Solution::gcd_of_strings(String::from("LEET"), String::from("CODE")),
        ""
    );
    assert_eq!(
        Solution::gcd_of_strings(String::from("ABCABC"), String::from("ABC")),
        "ABC"
    );
    assert_eq!(
        Solution::gcd_of_strings(String::from("ABABAB"), String::from("ABAB")),
        "AB"
    );
    assert_eq!(
        Solution::gcd_of_strings(String::from("ABABABAB"), String::from("ABAB")),
        "ABAB"
    );
    assert_eq!(
        Solution::gcd_of_strings(String::from("ABBAABBA"), String::from("ABBA")),
        "ABBA"
    );
    println!("All tests passed!")
}
